### PR TITLE
[RDY][CR] Refactor pouring of liquids + fix non-rigid containers + partly fix tool reloading

### LINKED ISSUE #16538: Small problem with exceeding volume limit while pouring liquid in resizable containers
Steps to reproduce:
1. Unload liquid into a sealed stomach that is in your inventory, so that the volume of the stomach would increase, and the inventory volume limit would be exceeded.
2. Try to put some items on a neighbouring cell. Get a error in the ingame-console and a message that some items tumble to the ground.

The items probably should tumble right after you unload liquid to the stomach, and the error should not appear.

### PR DESCRIPTION
**Pull request by codemime**
Fixes #16538.

EDIT:  ~~Marked WIP because the refactoring isn't finished yet.~~
### Summary:

Behavioral changes:
- Non-rigid containers can now adjust to the volume they're in (player's inventory) and don't exceed that volume [[proposed](https://github.com/CleverRaven/Cataclysm-DDA/issues/16538#issuecomment-217617924) by @kevingranade]. The game will explicitly tell it when a container has no room to expand
- Liquids can now be loaded-to / unloaded-from tools / guns that are charged by liquids (chainsaws, flamethrowers e.t.c) via liquid handling menu. This option was broken for some reason.
- When you pour liquid into an empty container it won't name the liquid being poured. So instead of e.g.`"You pour water into the plastic bottle of water"` the game will say `"You pour water into the plastic bottle"`.

Implementation changes:
- Simplification and de-duplication of the code
- Reloading is just a bit more unified with the gun reloading code. Moved a chunk of code from 'player::can_reload()' to 'item::is_reloadable_with()' where it clearly belongs. A known limitation: moves aren't consumed.
- Got rid of the liquid error codes (`L_ERR_` ...)
- `item::get_remaining_capacity_for_liquid()` can now take volume limitations into account
- `item::fill_with()` is now a pure modifier. It returns `void` and accepts only two arguments: `liquid` item and `amount`
- Added a sanity check into `item::fill_with()` to display a `debugmsg` when the container is not designed to contain liquids
- Replaced `item::reduce_charges()` with `item::mod_charges()`. It returns `void` and can add/remove charges. Handles infinite charges correctly now
- Introduced `item::liquid_charges_per_volume( int volume )` which says how many charges of liquid can be held in specified volume.

### CODE REVIEW

**Comment by codemime** [comment]
Does anybody know why does _сlang_ refuse to deal with a variadic lambda? _g++_ compiles it without a hitch.

**Comment by Coolthulhu** [comment]
Conflicts with #17351, but the conflict is pretty trivial.

**Comment by codemime** [comment]
I can resolve it if you're going to merge #17351 first.

**Comment by Coolthulhu** [comment]
Wait, this one seems to be unaffected, it must be the other liquid PR

File: src/item.cpp [hunk 1]
```diff
@@ -4955,66 +4978,58 @@ long item::get_container_capacity() const
...
-    if( !type->container->watertight ) {
-        return L_ERR_NOT_WATERTIGHT;
+        remaining_capacity = liquid.liquid_charges( get_container_capacity() );
+        if( !contents.empty() ) {
+            remaining_capacity -= contents.front().charges;
+        }
+        if( remaining_capacity > 0 && !type->rigid ) {
+            const long expansion = liquid.liquid_charges_per_volume( volume_limit );
+            if( expansion <= 0 ) {
+                return error( string_format( _( "That %s doesn't have a room to expand." ), tname().c_str() ) );
```
Comment (Coolthulhu) [inline src/item.cpp]: "a room" should drop the "a".

File: src/player.cpp [hunk 2]
```diff
@@ -8977,7 +8977,8 @@ item player::reduce_charges( int position, long quantity )
         debugmsg( "invalid item position %d for reduce_charges", position );
         return ret_null;
     }
-    if( it.reduce_charges( quantity ) ) {
+    it.mod_charges( -quantity );
```
Comment (Coolthulhu) [inline src/player.cpp]: Is there a guarantee that `reduce_charges` won't be invoked with `quantity > charges`?

Comment (codemime) [inline src/player.cpp]: No, but `mod_charges()` will show a `debugmsg` in that case (just like `item::reduce_charges()` did).

File: src/character.cpp [hunk 3]
```diff
@@ -1910,55 +1910,25 @@ bool Character::is_blind() const
...
-            return false;
-        }
-
-        if( container.ammo_remaining() && container.ammo_current() != liquid.typeId() ) {
-            add_msg_if_player( m_info, _( "You can't mix loads in your %s." ), container.tname().c_str() );
-            return false;
-        }
+    std::string err;
+
+    const bool allow_bucket = &container == &weapon || !has_item( container );
```
Comment (Coolthulhu) [inline src/character.cpp]: `has_item` is hacky. If it's the only non-minor hack here, I'd allow it, but still.
This should be passed from above where possible.

Comment (codemime) [inline src/character.cpp]: It's a [pre-existing hack](https://github.com/CleverRaven/Cataclysm-DDA/pull/17348/files#diff-f1933acfcc0056a11de5cb2ad6fd6caaL1946).

**Pull request merged**